### PR TITLE
refactor: LBWS specific ALB -> generic ALB

### PR DESCRIPTION
**Pull request by dannyrandall**
Refactors ALB code/templates currently specific to LBWS to be generic so it's easier to reuse in a backend service's internal alb. The main changes are:
- Rename parameter constants to be `Workload` constants instead of LBWS
- Pull ALB CF from LBWS CF into its own partial
- Move `manifest.RoutingRuleConfiguration` into it's own file from the LBWS file

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: internal/pkg/deploy/cloudformation/stack/backend_svc.go [hunk 1]
```diff
@@ -16,11 +16,6 @@ import (
 	"github.com/aws/copilot-cli/internal/pkg/template/override"
 )
 
-// Parameter logical IDs for a backend service.
-const (
-	BackendServiceContainerPortParamKey = "ContainerPort"
-)
```
Comment (iamhopaul123) [inline internal/pkg/deploy/cloudformation/stack/backend_svc.go:22]: 👍

File: internal/pkg/deploy/cloudformation/stack/lb_web_svc.go [hunk 2]
```diff
@@ -264,20 +258,16 @@ func (s *LoadBalancedWebService) Template() (string, error) {
...
-		targetContainer = s.manifest.RoutingRule.TargetContainer
-	}
-	if s.manifest.RoutingRule.TargetContainerCamelCase != nil {
-		targetContainer = s.manifest.RoutingRule.TargetContainerCamelCase
-	}
-	if aws.StringValue(targetContainer) != containerName {
+	targetContainer = aws.String(s.name)
+	targetPort = aws.String(s.containerPort())
+
+	rrTarget := s.manifest.RoutingRule.GetTargetContainer()
```
Comment (efekarakus) [inline internal/pkg/deploy/cloudformation/stack/lb_web_svc.go:265]: really nice 👌

**Review by iamhopaul123 — APPROVED** [review]
:shipit:

File: internal/pkg/manifest/http.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	DeregistrationDelay *time.Duration          `yaml:"deregistration_delay"`
+	// TargetContainer is the container load balancer routes traffic to.
+	TargetContainer          *string `yaml:"target_container"`
+	TargetContainerCamelCase *string `yaml:"targetContainer"` // "targetContainerCamelCase" for backwards compatibility
+	AllowedSourceIps         []IPNet `yaml:"allowed_source_ips"`
+}
+
+// GetTargetContainer returns the correct target container value, if set.
+// Use this function instead of getting r.TargetContainer or TargetContainerCamelCase directly.
+func (r *RoutingRuleConfiguration) GetTargetContainer() *string {
```
Comment (efekarakus) [inline internal/pkg/manifest/http.go:66]: just to validate, is this the only method added? everything else was just a mv operation?

**Comment by dannyrandall** [comment]
Adding `do-not-merge` label until https://github.com/aws/copilot-cli/pull/3548 goes through so I can wrap that into this refactor

File: internal/pkg/template/templates/workloads/partials/cf/alb.yml [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+              Effect: Allow
+              Action:
+                - "tag:GetResources"
+              Resource: "*"
+      {{- end}}
+    ManagedPolicyArns:
+      - !Sub arn:${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole
+
+{{- if .HTTPSListener}}
+{{include "https-listener" .}}
```
Comment (efekarakus) [inline internal/pkg/template/templates/workloads/partials/cf/alb.yml:110]: nice, i didn't know our recursive `include` functions worked 😂

**Review by efekarakus — APPROVED** [review]
LGTM too, just a question

File: internal/pkg/template/templates/workloads/partials/cf/alb.yml (hunk 4 above)
Comment (dannyrandall) [reply internal/pkg/template/templates/workloads/partials/cf/alb.yml:110]: apparently they do ¯\\\_(ツ)_/¯

File: internal/pkg/manifest/http.go (hunk 3 above)
Comment (dannyrandall) [reply internal/pkg/manifest/http.go:66]: yeah! and this function is effectively the same as the previous [`targetContainer()`](https://github.com/aws/copilot-cli/pull/3550/files#diff-8dc63309e718c405097ab0c6d168d527aa59e882bad9e0a0a144447c8e692a6dL258), just exported.

**Pull request merged**